packages/graphcache-rn-async-storage/src/makeAsyncStorage.ts

```diff
@@ -0,0 +1,114 @@
+import { StorageAdapter } from '@urql/exchange-graphcache';
+import AsyncStorage from '@react-native-async-storage/async-storage';
+import NetInfo from '@react-native-community/netinfo';
+
+export type StorageOptions = {
+  dataKey?: string;
+  metadataKey?: string;
+  maxAge?: number; // Number of days
+};
+
+let disconnect;
+
+export const makeAsyncStorage: (ops?: StorageOptions) => StorageAdapter = ({
+  dataKey = 'graphcache-data',
+  metadataKey = 'graphcache-metadata',
+  maxAge = 7,
+} = {}) => {
+  const todayDayStamp = Math.floor(
+    new Date().valueOf() / (1000 * 60 * 60 * 24)
+  );
+  const todayBatch = {};
+  const prefixCheck = new RegExp(`^${dataKey}_(\\d+)$`);
+
+  return {
+    /**
+     * On initial read, pull storage keys and see which match our storage key pattern.
+     * Any record that is expired should be removed from AsyncStorage.
+     * The rest of the batches should be merged into a cache in chrono order.
+     * Also hydrate todayBatch if we find it.
+     */
+    readData: async () => {
+      const cache = {};
+
+      try {
+        const persistedDayStamps = (await AsyncStorage.getAllKeys())
+          .reduce<number[]>((filtered, curr) => {
+            const match = curr.match(prefixCheck);
+            const ts = match ? match[1] : undefined;
+
+            if (ts) {
+              filtered.push(Number(ts));
+            }
+            return filtered;
+          }, [])
+          .sort();
+
+        // eslint-disable-next-line es5/no-for-of
+        for (const dayStamp of persistedDayStamps) {
+          // Discard
+          if (todayDayStamp - dayStamp > maxAge) {
+            await AsyncStorage.removeItem(`${dataKey}_${dayStamp}`);
+          }
+          // Parse batch and merge in.
+          else {
+            try {
+              const data = await AsyncStorage.getItem(`${dataKey}_${dayStamp}`);
+              if (data) {
+                const parsedData = JSON.parse(data);
+                Object.assign(cache, parsedData);
+
+                // We found today's batch, let's hydrate that while we're at it.
+                if (dayStamp === todayDayStamp) {
+                  Object.assign(todayBatch, parsedData);
+                }
+              }
+            } catch (_err) {}
```

Comment:
The await & try-catch combos read a bit unfortunate here and I'm not sure whether they're transpiled nicely. I'd opt for avoiding it or extracting it into a separate function. But overall, unwrapping into regular chained promises ends up much nicely in terms of transpiled code and readability in try-catch cases.

Reply:
I pulled out the async storage functions into utils, but the but had to [revert the change](https://github.com/FormidableLabs/urql/pull/1949/commits/0496e8f3c0263c74873b8d72deb535fd61f468a1) because the it caused the build command fail with:
```
Cannot read property 'constantViolations' of undefined
```
Looks like we can't define any async functions, e.g. adding
```js
async function anything() {
  return null;
}
```
will fail with the same error. I could write them using promises instead?